### PR TITLE
Support unneeded schema names in overviews queries

### LINKED ISSUE #420: Overviews not applied if unnecessary schema name is included in queries
If a layer's query references a table including the schema name, and the schema is not required given the `schema_path` in use (i.e. if the table regclass casted to text doesn't include the schema) overview tables are not applied to the query.

This is a problem because, for organisation users, the editor adds the schema part (user name) to
table references when the query SQL is edited manually.

### PR DESCRIPTION
**Pull request by jgoizueta**
Fixes #420

Keep table schema of overviews base tables and use it
to support queries that use the schema name when not
strictly needed.

### CODE REVIEW

**Comment by jgoizueta** [comment]
@rochoa @rafatower can you please review this?

**Comment by rafatower** [comment]
what you mean is that this makes the code independent of whether the schema is in the search path or not, right?

**Comment by rafatower** [comment]
uhm, no, what you mean is that the schema can be provided even if not needed (when it is already in the search path). Correct me if I'm wrong.

:+1:

**Comment by jgoizueta** [comment]
Yes, that's it: without this changes, if the schema is present when not needed (because it is in the search path and no other schema leading to a table with the same name is ahead of it in the path) overviews would not be used. Now overviews are applied in that case with or without having the explicit schema in the table name.

File: lib/cartodb/api/overviews_metadata_api.js [hunk 1]
```diff
@@ -31,11 +34,13 @@ OverviewsMetadataApi.prototype.getOverviewsMetadata = function (username, sql, c
         var metadata = {};
         rows.forEach(function(row) {
```
Comment (dgaubert) [inline lib/cartodb/api/overviews_metadata_api.js]: Can we use `reduce` approach here?. For instance:

```
var metadata = rows.reduce(function (metadata, row) {
    metadata = ....
    ...
    return metadata;
}, {});
```

File: test/unit/cartodb/overviews_query_rewriter.js [hunk 2]
```diff
@@ -202,6 +202,78 @@ describe('Overviews query rewriter', function() {
     done();
   });
 
+  it('uses schema name from overviews', function(done){
```
Comment (dgaubert) [inline test/unit/cartodb/overviews_query_rewriter.js]: It seems a synchronous test, in that case, `done`callback is not needed.

**Comment by dgaubert** [comment]
if you can fix two spaces indentation would be nice! ;)

File: lib/cartodb/api/overviews_metadata_api.js (hunk 1 above)
Comment (jgoizueta) [inline lib/cartodb/api/overviews_metadata_api.js]: sure thing

**Comment by jgoizueta** [comment]
Man there was a ton of bad indentation! :sweat:

So, after the next comment the overall diff will be useless...

**Comment by dgaubert** [comment]
Ok, don't do it then! We will fix later then ;)

**Comment by jgoizueta** [comment]
just did it!

File: lib/cartodb/api/overviews_metadata_api.js [hunk 3]
```diff
@@ -22,22 +22,26 @@ function prepareSql(sql) {
...
         if (err){
             callback(err);
             return;
         }
-        var metadata = {};
-        rows.forEach(function(row) {
+        var metadata = rows.reduce(function(metadata, row){
             var table = row.base_table;
+            var schema = row._cdb_schema_name;
             var table_metadata = metadata[table];
```
Comment (dgaubert) [inline lib/cartodb/api/overviews_metadata_api.js]: I'm not sure what `table_metadata` does. I mean, it is only used to check `metadata[table]`, right?. I think we can remove it!

Comment (jgoizueta) [inline lib/cartodb/api/overviews_metadata_api.js]: Right, can be removed, I'm on it

**Comment by dgaubert** [comment]
:+1:

**Pull request merged**